src/rootfstreemanager.cc

```diff
@@ -250,3 +263,134 @@ data::InstallationResult RootfsTreeManager::verifyBootloaderUpdate(const Uptane:
+      LOG_ERROR << "No delta stat found between " << getCurrentHash() << " and " << target.Sha256Hash()
+                << "; skipping update size checking";
+      return true;
+    }
+    // TODO: watermark and add stat to the update even context
+    return canDeltaFitOnDisk(delta_stat);
+  } catch (const std::exception& exc) {
+    LOG_ERROR << "Failed to check if the update can fit on a disk: " << exc.what();
+  }
+  return res;
+}
+
+bool RootfsTreeManager::canDeltaFitOnDisk(const DeltaStat& delta_stat) const {
+  const uint64_t free_space{getFreeSpace(sysroot_->path())};
+  // TODO: use storage watermark
+  // TODO: check if `delta_stat.size` compressed size should be taken into account too
+  return delta_stat.uncompressedSize < free_space;
+}
+
+bool RootfsTreeManager::getDeltaStatsRef(const Json::Value& json, DeltaStatsRef& ref) {
+  if (!json.isMember("delta-stats")) {
+    return false;
+  }
+  const auto delta_stats_ref{json["delta-stats"]};
+  if (!delta_stats_ref.isMember("sha256") || !delta_stats_ref["sha256"].isString()) {
+    LOG_ERROR << "Incorrect metadata about static delta statistics are found in Target;"
+              << " err: missing `sha256` field or it's not a string";
+    return false;
+  }
+  if (!delta_stats_ref.isMember("size") || !delta_stats_ref["size"].isUInt()) {
+    LOG_ERROR << "Incorrect metadata about static delta statistics are found in Target;"
+              << " err: missing `size` field or it's not an integer";
+    return false;
+  }
+  ref = {delta_stats_ref["sha256"].asString(), delta_stats_ref["size"].asUInt()};
+  return true;
+}
+
+Json::Value RootfsTreeManager::downloadDeltaStats(const DeltaStatsRef& ref, const Remote& remote) {
+  static const uint64_t DeltaStatsMaxSize{1024 * 1024};
+  const std::string uri = remote.baseUrl + "/delta-stats/" + ref.sha256;
+
+  if (ref.size > DeltaStatsMaxSize) {
+    LOG_ERROR << "Requested delta stat file has higher size than maximum allowed; "
+                 " requested size: "
+              << ref.size << ", maximum allowed: " << DeltaStatsMaxSize;
+    return Json::nullValue;
+  }
+  std::vector<std::string> extra_headers;
+  for (const auto& header : remote.headers) {
+    extra_headers.emplace_back(header.first + ": " + header.second);
+  }
+  auto client{HttpClient(&extra_headers)};
+
+  LOG_INFO << "Fetching delta stats -> " << uri;
+  const auto resp = client.get(uri, ref.size);
+  if (!resp.isOk()) {
+    LOG_ERROR << "Failed to fetch static delta stats; status: " << resp.getStatusStr() << ", err: " << resp.body;
+    return Json::nullValue;
+  }
+  if (resp.body.size() != ref.size) {
+    LOG_ERROR << "Fetched invalid static delta stats, size mismatch; "
+              << " expected: " << ref.size << ", got: " << resp.body.size();
+    return Json::nullValue;
+  }
+  const auto received_data_hash{
+      boost::algorithm::to_lower_copy(boost::algorithm::hex(Crypto::sha256digest(resp.body)))};
```

Comment:
so we require canonicaljson as our input, correct?

Reply:
Yes.